src/main.c

```diff
@@ -1565,13 +1565,16 @@ static void on_accept(h2o_socket_t *listener, const char *err)
 
     do {
         h2o_socket_t *sock;
-        if (num_connections(0) >= conf.max_connections) {
+        int excess_connections = num_connections(0) - conf.max_connections;
+        if (excess_connections >= 0) {
             /* The accepting socket is disactivated before entering the next in `run_loop`.
              * Note: it is possible that the server would accept at most `max_connections + num_threads` connections, since the
              * server does not check if the number of connections has exceeded _after_ epoll notifies of a new connection _but_
              * _before_ calling `accept`.  In other words t/40max-connections.t may fail.
              */
-            break;
+            if (!h2o_context_close_idle_connections(ctx->accept_ctx.ctx, excess_connections)) {
```

Comment:
I still see two issues in this line:

1. we should pass `excess_connections + 1` to ensure that there will be at least one slot for the incoming connection
2. for now `h2o_context_close_idle_connections` returns the number of closed connections, so we should compare that with `excess_connections + 1`

Thus I think the following is the correct one:
```c
if (h2o_context_close_idle_connections(ctx->accept_ctx.ctx, excess_connections + 1) < excess_connections + 1) {
```

Reply:
And also I realized that it's dangerous to pass `0` to `h2o_context_close_idle_connections` mistakenly like this case, because it will try to close all idle connections if we do so. How about adding `assert(max_connections_to_close > 0)` in the beginning of that function?